rpc/src/v1/helpers/nonce.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+use futures::{Future, Poll, Async};
+use util::{U256, RwLock};
+use std::collections::{VecDeque};
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum NonceError {
+    Dropped,
+    InvalidPoll,
+}
+
+#[derive(Copy, Clone)]
+enum NonceState {
+    Prospective,
+    Reserved,
+    Dispatch,
+    Error(NonceError),
+}
+
+pub struct Nonce {
+    state: NonceState,
+    value: U256,
+    reserved: Arc<RwLock<VecDeque<U256>>>,
+}
+
+impl Drop for Nonce {
+    fn drop(&mut self) {
+        match self.state {
+            NonceState::Reserved => {
+                let reserved = self.reserved.read();
+                if let Some(idx) = reserved.iter().position(|x| *x == self.value) {
+                    drop(reserved);
+                    let mut reserved = self.reserved.write();
+                    reserved.drain(idx..); // any reservations after this one are now invalid
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+impl Future for Nonce {
+    type Item = U256;
+    type Error = NonceError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self.state {
+            NonceState::Prospective => {
+                let result = if let Some(reserved) = self.reserved.try_read() {
+                    while reserved.contains(&self.value) { // reserved by another thread
+                        self.value = self.value + U256::one(); // increment prospective nonce
+                    }
+                    drop(reserved);
+                    // prospective nonce is available
+                    if let Some(mut reserved) = self.reserved.try_write() {
+                        reserved.push_back(self.value);
+                        // use this nonce for signing
+                        // then poll for dispatch readiness
+                        // dispatch may fail if another thread drops its nonce
+                        Ok(Async::Ready(self.value))
+                    } else {
+                        Ok(Async::NotReady)
```

Comment:
I don't think it's valid to return `Async::NotReady` in case the lock is not available. I think it might never be polled again since `RwLock` won't notify the `Task` that the state changed.